In [2]:
import pandas as pd
import numpy as np
from tqdm.auto import tqdm
import sys
sys.path.append("../oracle-polimi-contest-2019/")
from evaluation_script import *

In [3]:
score_path = '../lgb_predictions_full.csv'
s = pd.read_csv(score_path)

In [4]:
train = pd.read_csv("../dataset/original/train.csv", escapechar="\\")
test = pd.read_csv("../dataset/original/test.csv", escapechar="\\")

In [5]:
s.ordered_scores = [eval(x) for x in s.ordered_scores]
s.ordered_linked = [eval(x) for x in s.ordered_linked]
s.ordered_record = [eval(x) for x in s.ordered_record]

In [6]:
def expand_df(s):
    # Expand df
    new_df = []
    for (q, l, r, sc) in tqdm(zip(s.queried_record_id, s.ordered_linked, s.ordered_record, s.ordered_scores)):
            for x in range(len(l)):
                new_df.append((q, l[x], r[x], sc[x]))

    sub = pd.DataFrame(new_df, columns=['queried_record_id', 'predicted_linked_id', 'predicted_record_id', 'score'])
    return sub

In [6]:
sub = expand_df(s)

In [7]:
s

queried_record_id  \
0           10000003-TST-MR   
1            10000008-TST-M   
2           10000010-TST-CP   
3           10000013-TST-MR   
4           10000016-TST-MR   
5           10000017-TST-MR   
6        10000018-T4-TST-CP   
7            10000018-TST-M   
8       10000020-NV0-TST-CP   
9        10000020-NV1-TST-M   
10          10000022-TST-CP   
11      10000023-NV2-TST-CP   
12          10000024-TST-MR   
13          10000025-TST-MR   
14        10000027-M0-TST-M   
15        10000027-T0-TST-M   
16        10000027-T1-TST-M   
17          10000028-TST-MR   
18          10000030-TST-MR   
19        10000031-M2-TST-M   
20       10000032-T0-TST-CP   
21        10000032-T1-TST-M   
22           10000034-TST-M   
23       10000036-NV0-TST-M   
24       10000036-NV1-TST-M   
25       10000036-T0-TST-CP   
26           10000036-TST-M   
27          10000037-TST-MR   
28          10000038-TST-MR   
29          10000039-TST-CP   
...                     ...   
266925       13013118-TST-M   
266926  13013119-NV0-TST-CP   
266927       15000601-TST-M   
266928      15000801-TST-MR   
266929      15000901-TST-MR   
266930         15001201-NV0   
266931      15001301-TST-MR   
266932   15001401-M0-TST-CP   
266933   15001401-M2-TST-CP   
266934      15001501-TST-MR   
266935   15001601-T3-TST-CP   
266936   15001801-T0-TST-CP   
266937      15001903-TST-MR   
266938    15002501-M0-TST-M   
266939      15003201-TST-MR   
266940      15003301-TST-MR   
266941   15004001-T0-TST-CP   
266942    15004001-T3-TST-M   
266943      15004501-TST-MR   
266944      15005002-TST-MR   
266945      15005003-TST-MR   
266946       15005004-TST-M   
266947      15005201-TST-MR   
266948      15005401-TST-MR   
266949       15005501-TST-M   
266950    15005801-T0-TST-M   
266951    15005801-T1-TST-M   
266952      15006601-TST-MR   
266953      15006701-TST-CP   
266954      15006801-TST-MR   

                                                 rec_pred  \
0       [(10010930, -9.196762530524452, '10010930-NV0'...   
1       [(10043172, 1.726297182553191, '10043172-NV0')...   
2       [(10000010, 7.7471028268421875, '10000010'), (...   
3       [(10153043, -6.795391757680415, '10153043'), (...   
4       [(10104833, -5.524791284191256, '10104833'), (...   
5       [(10094323, -3.6202485155222233, '10094323-T0'...   
6       [(10000018, 8.248854345239275, '10000018-T4'),...   
7       [(10000018, 8.248854345239275, '10000018-T1'),...   
8       [(10000020, 3.25426347945908, '10000020'), (10...   
9       [(10000020, 7.588364448706042, '10000020'), (1...   
10      [(10000022, 7.918333164537388, '10000022'), (1...   
11      [(10000023, 7.522166758783997, '10000023'), (1...   
12      [(10100649, 3.398486141873055, '10100649'), (1...   
13      [(10038568, 4.440899798772405, '10038568'), (1...   
14      [(10000027, 7.774715643577811, '10000027'), (1...   
15      [(10000027, 7.774715643577811, '10000027'), (1...   
16      [(10000027, 7.524012323479619, '10000027'), (1...   
17      [(10074750, -9.079772698405694, '10074750'), (...   
18      [(10037339, 3.55973801898433, '10037339'), (10...   
19      [(10000031, 1.337676613248445, '10000031'), (1...   
20      [(10000032, 7.850064053878422, '10000032-T0'),...   
21      [(10000032, 7.169220156482475, '10000032'), (1...   
22      [(10119069, -1.3138606494309073, '10119069-T0'...   
23      [(10000036, 7.812863548620348, '10000036-T0'),...   
24      [(10000036, 7.677987582087094, '10000036-T0'),...   
25      [(10000036, 7.812863548620348, '10000036-T0'),...   
26      [(10000036, 7.812863548620348, '10000036-T0'),...   
27      [(10120582, -5.21671912871256, '10120582'), (1...   
28      [(10136074, -5.298384479275279, '10136074-NV0'...   
29      [(10000039, 7.906224249400227, '10000039'), (1...   
...                                                   ...   
266925  [(13013118, 3.9727005475287926, '13013118-NV0'...   
266926  [(13013119, 7.603385123002431, '13013119-NV0')...   
266927  [(130

# Add linked_id to the test

In [8]:
test['linked_id'] = test.record_id.str.split("-")
test['linked_id'] = test.linked_id.apply(lambda x: x[0])
test.linked_id = test.linked_id.astype(int)

# Top10

In [9]:
precision_at_k(sub[['queried_record_id', 'predicted_record_id']], train.set_index("record_id"), test.set_index("record_id"))

processed 1000/266955 records, 0.37%
	current precision: 0.45
	time elapsed: 12.96 s
processed 2000/266955 records, 0.75%
	current precision: 0.46
	time elapsed: 13.44 s
processed 3000/266955 records, 1.12%
	current precision: 0.46
	time elapsed: 13.94 s
processed 4000/266955 records, 1.50%
	current precision: 0.46
	time elapsed: 14.41 s
processed 5000/266955 records, 1.87%
	current precision: 0.47
	time elapsed: 14.92 s
processed 6000/266955 records, 2.25%
	current precision: 0.46
	time elapsed: 15.40 s
processed 7000/266955 records, 2.62%
	current precision: 0.47
	time elapsed: 15.86 s
processed 8000/266955 records, 3.00%
	current precision: 0.47
	time elapsed: 16.34 s
processed 9000/266955 records, 3.37%
	current precision: 0.47
	time elapsed: 16.80 s


KeyboardInterrupt: 

In [ ]:
recall_at_k(sub[['queried_record_id', 'predicted_record_id']], train.set_index("record_id"), test.set_index("record_id"))

# Restricted

La precision è calcolata in questo modo:
### precision = len(selectedRelevantRecords) / len(setPredictedRecords)
Riducendo il numero di predizioni a quelle 'sicure', dovremmo aumentare la precision stessa. Quello che è stato fatto è questo: supponiamo di avere

| index | queried_record_id |predicted_linked_id | predicted_record_id | score
| --- | --- | --- | --- | --- |
| 0 | 15004001-T0-TST-CP | 15004001 | 15004001-T0 | 7.553092
| 1 | 15004001-T0-TST-CP | 15004001 | 15004001-M0 | 7.553092
| 2 | 15004001-T0-TST-CP |	15004001 |	15004001 |	7.553092
| 3 |	15004001-T0-TST-CP |	15004001 |	15004001-T2 |	7.553092
| 4 |	15004001-T0-TST-CP |	15004001 |	15004001-T1 |	4.449898
| 5 |	15004001-T0-TST-CP |	15004001 |	15004001-T4 |	4.403244
| 6 |	15004001-T0-TST-CP |	15004001 |	15004001-M2 |	1.808847
| 7 |	15004001-T0-TST-CP |	15004001 |	15004001-M1 |	1.808847
| 8 |	15004001-T0-TST-CP |	15004001 |	15004001-NV1 |	0.667029
| 9 |	15004001-T0-TST-CP |	12187297 |	12187297 |	-7.709455

il **delta** corrisponde alla differenza tra lo score più alto e l'i-esimo score: s[0] - s[i]. Per definire una certa 'sicurezza' nelle raccomandazioni poniamo max_delta = 2.0. Dunque la submission ristretta considera, tra i top10, quelli che hanno score distante <= max_delta, inoltre vengono comunque inserite quelle predizioni k tra le top10 che hanno lo stesso linked_id della predizione con score più alto (il primo): linked_id_k == linked_id_0. In questo modo raccomandiamo i record_id di cui siamo 'sicuri'. Nell'esempio, le row 1,2 e 3 vengono mantenute come raccomandazioni poiché hanno lo score che dista <= di max_delta. Inoltre le row 4,5,6,7 e 8 vengono comunque mantenute poiché hanno lo stesso linked_id della row 0. L'unico che viene escluso è la row 9 perché lo score > max_delta e il linked_id non è lo stesso.

In [ ]:
def restricted_df(s):
    restricted_pred = []
    max_delta = 2.0
    for (q, sc, rec, l) in tqdm(zip(s.queried_record_id, s.ordered_scores, s.ordered_record, s.ordered_linked)):
        for x in range(len(sc)):
            if x == 0: # Il primo elemento predetto si mette sempre [quello con score più alto]
                restricted_pred.append((q, sc[x], rec[x]))
            else:
                if x >= 10:
                    continue
                elif (sc[0] - sc[x] < max_delta) or (l[0] == l[x]):   # se le predizioni hanno uno scores che dista < max_delta dalla prima allora si inseriscono
                    restricted_pred.append((q, sc[x], rec[x]))
                else:
                    continue
    restricted_df = pd.DataFrame(restricted_pred, columns=['queried_record_id', 'scores', 'predicted_record_id'])
    return restricted_df

In [ ]:
restricted_df = restricted_df(s)

In [ ]:
precision_at_k(restricted_df[['queried_record_id', 'predicted_record_id']], train.set_index("record_id"), test.set_index("record_id"))

In [ ]:
recall_at_k(restricted_df[['queried_record_id', 'predicted_record_id']], train.set_index("record_id"), test.set_index("record_id"))

# F1-score

In [ ]:
# TODO calcolare l'F1 score, la formula è questa
#F1 = 2 * (precision * recall) / (precision + recall)

In [1]:
def compute_f1_score(precision, recall):
    F1 = 2 * (precision * recall) / (precision + recall)
    return F1

## Cosa sbagliamo? Mettere nell'articolo ciò che sbagliamo in modo da far capire che in alcuni casi possiamo fare poco o nulla

In [100]:
train['phone'] = train.phone.astype(str)
test['phone'] = test.phone.astype(str)

In [10]:
def isin_function(pred, linked):
    res = np.empty(len(linked), dtype=int)
    for i in tqdm(range(len(linked))):
        res[i] = np.isin(linked[i], pred[i])
    return res

In [15]:
# Get only first 10 predictions
predicted_linked = s.ordered_linked.values
predicted_linked = [x[:10] for x in predicted_linked]

In [26]:
test = test.sort_values(by='record_id').reset_index(drop=True)

In [27]:
s['correct'] = isin_function(predicted_linked, test.linked_id.values) 

In [34]:
# do not consider queried records that are only in test
train['linked_id'] = train.linked_id.astype(int)
only_test = set(test.linked_id) - set(train.linked_id)
only_test_rec = test[test.linked_id.isin(only_test)].record_id.values

In [39]:
wrong_pred = s[~s.queried_record_id.isin(only_test_rec) & (s.correct == 0)]

In [136]:
wrong_pred_rec = iter(wrong_pred.queried_record_id.values)

In [185]:
curr = next(wrong_pred_rec)

In [186]:
test[test.record_id == curr]

record_id                  name    type address         phone  \
16889  10024343-T2-TST-M  FAyHOy fOLDINGS LTD.  entity     NaN  3910990359.0   

      email modification  linked_id  
16889   NaN     move row   10024343

In [187]:
#real 
l_test = test[test.record_id == curr].linked_id.values[0]
train[train.linked_id == l_test]

record_id                  name    type address         phone email  \
34427      10024343  FATHOM HOLDINGS LTD.  entity     NaN  3910990359.0   NaN   
187108  10024343-M1  FATHOM HOLDINGS LTD.  entity     NaN           nan   NaN   
383407  10024343-T3  FATHOM HOLDINGStLTD.  entity     NaN  3910990359.0   NaN   
384044  10024343-T0  FATHOM HOLDINGS LTD.  entity     NaN  3910990359.0   NaN   
451269  10024343-T1  FATHOM HOLDINGS LTD.  entity     NaN  3910990359.0   NaN   

       modification  linked_id  
34427      original   10024343  
187108      missing   10024343  
383407         typo   10024343  
384044         typo   10024343  
451269         typo   10024343

In [188]:
first_pred = s[s.queried_record_id == curr].ordered_record.values[0][0]

In [189]:
# predicted
train[train.record_id == first_pred]

record_id               name    type address phone email  \
666412  10179670-M0  YAP HOLDINGS LTD.  entity     NaN   nan   NaN   

       modification  linked_id  
666412      missing   10179670

In [190]:
s[s.queried_record_id == curr].ordered_scores.values[0][0]

-4.941296771187973

In [ ]:
# casi importanti: 10014653-T2-TST-M, 

In [111]:
a = s[s.queried_record_id == curr].ordered_record.values[0]

In [113]:
dfa = train[train.record_id.isin(a)]

In [114]:
dfa

record_id                           name          type  \
24142    12202587-T2                    CHU CkIxSHI       officer   
44845    10056352-T1     aYAN INTERNATIONAL LIMITED        entity   
58913    13004941-T1                    LI Ckun Yui       officer   
62975       12162921            John Michael Gaynor       officer   
64762   10056352-NV2     RYAN INTERNATIONAL LIMITED        entity   
80078    10168656-T3                   Calaff Cojp.        entity   
85185   10180512-NV0                ALKO IMPEX LTD.        entity   
93876    12112459-T1                   GARY DEMETRI       officer   
99809    10082049-T2            CORMACK CkrPORATION        entity   
101204   11006616-M1              BKS FAMILY OFFICE  intermediary   
123974  12154029-NV2      Silvia Adriana De Angelis       officer   
161532      12101603                GARY ABRAHAMSON       officer   
184537   13006924-T4                 LIf BOOw CkUAN       officer   
189698   13006002-T0                LOq SjANG CkUAN       officer   
196892   12122865-T0               MICaAEf C REDMAN       officer   
248145  10147105-NV0    Morze International Limited        entity   
255480      12062137     PERCY SAMUEL URIARTE OTOYA       officer   
256522   10040931-T0                  SpIVARf CORP.        entity   
320334   11008798-T2               rR CkRLO STEFANI  intermediary   
341146   12035281-T1               CHANj JUNn CkIEH       officer   
354565      12007171  TAUNUS INVESTMENTS GROUP S.A.       officer   
370006      10055380          SWISS CAPITAL LIMITED        entity   
378172      10168310       BRADE PROPERTIES LIMITED        entity   
384826   12025339-T0               cR. CLAkDf COHEN       officer   
398698   10196656-T1                MARVESTOf CORPa        entity   
405004   12106178-T0                  WuNG CkUANLIN       officer   
408955   10137390-T1          rST Breef Ckrporatiun        entity   
415227   12221167-T2                 JEN CkaNG LIzG       officer   
444727   10168310-T1       BRADE PROPERTIES LIMITED        entity   
457260  10056352-NV1     RYAN INTERNATIONAL LIMITED        entity   
461345   12150385-T2                    NG CkING YU       officer   
463343      12156938        First Names One Limited       officer   
466954      11003263               MOSSFON MANAGERS  intermediary   
468639   10138823-T3                   Llnqdi Corm.        entity   
489131   12183116-T4                    Mt CkAO JjE       officer   
493630      10056352     RYAN INTERNATIONAL LIMITED        entity   
498571   10023937-T2                  RAMBURY CkRP.        entity   
504316   13010893-M0                  Jozef Cukurel       officer   
508931  10147105-NV1    Morze International Limited        entity   
515392   10049852-T2      LIQUITART PROPERTIES INC.        entity   
546302   12105420-T3       MINERVA NOMINEES LIMITED       officer   
555099  10009001-NV0       VOYAGER INVESTMENTS S.A.        entity   
561789   10056352-T0     RYAN INTERNATIONALdLIMITED        entity   
568806   13010893-T0                  Jowef Cukurel       officer   
575798   12101603-M0                GARY ABRAHAMSON       officer   
578337   10205790-T0        MARSHALL LINK Ck., oTm.        entity   
580433   13010893-M2                  Jozef Cukurel       officer   
584579   12201290-T1                 LIf CHANGscING       officer   
590331      12045562                         BEARER       officer   
595747      12112459                   GARY DEMETRI       officer   
595847      12004417                   GAO ZHU QING       officer   
603378   10084269-T0                JAS CkrPORrTION        entity   
610102   10081067-T0                   JtELRf CsRP.        entity   
613580  13012611-NV0                MOHAMMAD ZAHOOR       officer   
637630   10167201-T0               Lastn Gyouf Corp        entity   
642077   10012090-T0               WINDSTER LIMIwED        entity   
645158      13010893                  Jozef Cukurel       officer   
659373  1